### PR TITLE
DHFPROD-6708: Save mappings to related entity types in Hub Central

### PR DESCRIPTION
**Pull request by briantang**
### Description
- Includes [DHFPROD-7193: Add an entity-specific context field](https://project.marklogic.com/jira/browse/DHFPROD-7193) 
- Tests added in RTL
- cypress tests will be needed but require extensive setup to cover related entities from end-to-end. Created a task for that [here](https://project.marklogic.com/jira/browse/DHFPROD-7267)
#### Checklist: 
```diff
- Note: do not change the below
```

-  ##### Owner:

- [x] JIRA_ID included in all the commit messages
- [x] PR title is in the format JIRA_ID:Title
- [x] Rebase the branch with upstream
- [x] Squashed all commits into a single commit
- [x] Code passes ESLint tests
- [x] Added Tests
  

- ##### Reviewer:

- [x] Reviewed Tests

### CODE REVIEW

**Review by timur-isangulov — CHANGES_REQUESTED** [review]
@briantang, I noticed main entity previously saved mappings disappear after saving mappings
Screen Shot 2021-04-26 at 8 07 22 AM
 for the related entity.

File: examples/reference-entity-model/steps/mapping/mapCustomersWithRelatedEntitiesJSON.step.json [hunk 1]
```diff
@@ -98,7 +98,6 @@
           "sourcedFrom": "SendToID"
         },
         "lineItems": {
-          "sourcedFrom": "Products",
```
Comment (srinathgit) [inline examples/reference-entity-model/steps/mapping/mapCustomersWithRelatedEntitiesJSON.step.json:101]: Why is this removed ?

**Review by srinathgit — CHANGES_REQUESTED** [review]
1. When I open and the mapping has related entity mappings, they don't show up by default
2. It becomes extremely difficult to select for mappings that have few properties. And also the search box is misplaced (see attachment)
Screen Shot 2021-04-26 at 8 47 55 AM

File: examples/reference-entity-model/steps/mapping/mapCustomersWithRelatedEntitiesJSON.step.json (hunk 1 above)
Comment (briantang) [reply examples/reference-entity-model/steps/mapping/mapCustomersWithRelatedEntitiesJSON.step.json:101]: Thanks @srinathgit . That line should not have been removed. Will make the change

**Comment by briantang** [comment]
> @briantang, I noticed that main entity mappings disappear after saving mappings for the related entity.
> Screen Shot 2021-04-26 at 8 07 22 AM

Thanks @timur-isangulov, pushed a fix now.

**Comment by briantang** [comment]
> 1. When I open and the mapping has related entity mappings, they don't show up by default
> 2. It becomes extremely difficult to select for mappings that have few properties. And also the search box is misplaced (see attachment)
> 
> Screen Shot 2021-04-26 at 8 47 55 AM

Thanks @srinathgit, discussed with PM and will be branching these two issues out into bugs for next sprint. The other issue with the Xpath context is fixed now.

**Review by timur-isangulov — APPROVED** [review]
Looks good.

**Review by wooldridge — CHANGES_REQUESTED** [review]
I discovered the following issue:

1. With reference project, create a new mapping (MapTest).
2. Use collection: loadCustomersWithRelatedEntitiesJSON
3. Open the mapping details. Set some XPath Expression values for Customer.
4. Open the zip structured property under Customer billing to show its subproperties.
5. Insert a value into a zip subproperty using the property menu.
6. The mapping is saved, but the zip structure property is then automatically closed.

This autoclosure doesn't occur with subproperties one level up from zip (e.g., shipping, billing).

I expect the automatic closing not to happen.

Screencast: https://drive.google.com/file/d/13878WIJjAsTS7sKUF0HIUzvKlovtfxw3/view

**Comment by briantang** [comment]
> I discovered the following issue:
> 
> 1. With reference project, create a new mapping (MapTest).
> 2. Use collection: loadCustomersWithRelatedEntitiesJSON
> 3. Open the mapping details. Set some XPath Expression values for Customer.
> 4. Open the zip structured property under Customer billing to show its subproperties.
> 5. Insert a value into a zip subproperty using the property menu.
> 6. The mapping is saved, but the zip structure property is then automatically closed.
> 
> This autoclosure doesn't occur with subproperties one level up from zip (e.g., shipping, billing).
> 
> I expect the automatic closing not to happen.
> 
> Screencast: https://drive.google.com/file/d/13878WIJjAsTS7sKUF0HIUzvKlovtfxw3/view

Thanks @wooldridge, this is fixed now.

File: marklogic-data-hub-central/ui/src/components/entities/mapping/mapping-step-detail/mapping-step-detail.tsx [hunk 2]
```diff
@@ -478,19 +475,22 @@ const MappingStepDetail: React.FC = () => {
...
             let relatedEntProps = entityObject["entityModel"].definitions[relatedEntityName].properties;
             let relatedEntityTempData: any =[];
+            let contextKey = EntityTableKeyIndex + 1;
+            EntityTableKeyIndex++;
             let relatedEntityProps = extractNestedEntityData(relatedEntProps, relatedEntityTempData);
-            setRelatedEntityTypeProperties(prevState => ([...prevState, {entityType: entityObject.entityType, entityLabel: entityObject.mappingTitle, entityMappingId: entityObject.entityMappingId, relatedEntityMappings: entityObject.relatedEntityMappings, entityProps: relatedEntityProps}]));
+            relatedEntityProps.unshift({key: contextKey, name: "Context", type: "", isProperty: false, filterName: "Context", filterMatch: false}); //add Context field to front of properties
+            setRelatedEntityTypeProperties(prevState => ([...prevState, {entityType: entityObject.entityType, entityModel: entityObject.entityModel, entityLabel: entityObject.mappingTitle, entityMappingId: entityObject.entityMappingId, relatedEntityMappings: entityObject.relatedEntityMappings, entityProps: relatedEntityProps}]));
```
Comment (timur-isangulov) [inline marklogic-data-hub-central/ui/src/components/entities/mapping/mapping-step-detail/mapping-step-detail.tsx:489]: This line sometimes cause duplication of the entity objects in the relatedEntityTypeProperties array. 
Screen Shot 2021-04-27 at 11 44 23 AM

Comment (briantang) [reply marklogic-data-hub-central/ui/src/components/entities/mapping/mapping-step-detail/mapping-step-detail.tsx]: Thanks @timur-isangulov for the catch and assistance. Pushed a fix.

**Review by timur-isangulov — APPROVED** [review]

**Review by wooldridge — DISMISSED** [review]

**Review by srinathgit — CHANGES_REQUESTED** [review]
Everything looks good. One minor change, anytime a related entity mapping is added, we are making a call to /api/artifacts/mapping/functions which may not be necessary

**Comment by briantang** [comment]
> Everything looks good. One minor change, anytime a related entity mapping is added, we are making a call to /api/artifacts/mapping/functions which may not be necessary

Good point @srinathgit, I've removed the redundant api call and pushed. Ready for re-review

**Review by srinathgit — APPROVED** [review]

**Review by timur-isangulov — APPROVED** [review]
